go/pkg/services/logcollector/server.go

```diff
@@ -0,0 +1,605 @@
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	// read log from file
+	buffer, err := s.readLogsFromFile(ctx, request.RunUID, filePath, request.Offset, request.Size)
+	if err != nil {
+		err := errors.Wrapf(err, "Failed to read logs for run id %s", request.RunUID)
+		return &protologcollector.GetLogsResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	return &protologcollector.GetLogsResponse{
+		Success: true,
+		Logs:    buffer,
+	}, nil
+}
+
+// startLogStreaming streams logs from a pod and writes them into a file
+func (s *LogCollectorServer) startLogStreaming(ctx context.Context,
+	runUID,
+	podName string,
+	startedStreamingGoroutine chan bool,
+	cancelCtxFunc context.CancelFunc) {
+
+	// in case of a panic, remove this goroutine from the in-memory state, so the
+	// monitoring loop will start logging again for this runUID.
+	defer func() {
+		if err := recover(); err != nil {
+			callStack := debug.Stack()
+			s.Logger.ErrorWithCtx(ctx, "Panic caught while creating function",
+				"err", err,
+				"stack", string(callStack))
+
+			// remove this goroutine from in-memory state
+			if err := s.inMemoryState.RemoveLogItem(runUID); err != nil {
+				s.Logger.WarnWithCtx(ctx, "Failed to remove item from in memory state")
+			}
+
+			// cancel all other goroutines spawned from this one
+			cancelCtxFunc()
+		}
+	}()
+
+	// signal "main" function that goroutine is up
+	startedStreamingGoroutine <- true
+
+	// create a log file to the pod
+	logFilePath := s.resolvePodLogFilePath(runUID, podName)
+	if err := common.EnsureFileExists(logFilePath); err != nil {
+		s.Logger.ErrorWithCtx(ctx,
+			"Failed to ensure log file",
+			"runUID", runUID,
+			"logFilePath", logFilePath)
+		return
+	}
+
+	// get logs from pod, and keep the stream open (follow)
+	podLogOptions := &v1.PodLogOptions{
+		Follow: true,
+	}
+	restClientRequest := s.kubeClientSet.CoreV1().Pods(s.namespace).GetLogs(podName, podLogOptions)
+
+	// initialize stream and error for the while loop
+	var (
+		streamErr error
+		stream    io.ReadCloser
+	)
+
+	// stream logs - retry if failed
+	for {
+		stream, streamErr = restClientRequest.Stream(ctx)
+		if streamErr == nil {
+			break
+		}
+		s.Logger.WarnWithCtx(ctx, "Failed to get pod log stream read/closer, retrying", "runUID", runUID, "err", streamErr.Error())
+		time.Sleep(1 * time.Second)
+	}
+	defer stream.Close() // nolint: errcheck
+
+	for {
+		keepLogging, err := s.streamPodLogs(ctx, runUID, logFilePath, stream)
+		if err != nil {
+			s.Logger.WarnWithCtx(ctx, "An error occurred while streaming pod logs", "err", err)
+		}
+		if keepLogging {
+			continue
+		}
+		break
+	}
+
+	s.Logger.DebugWithCtx(ctx,
+		"Removing item from state file",
+		"runUID", runUID,
+		"podName", podName)
+
+	// remove run from state file
+	if err := s.stateStore.RemoveLogItem(runUID); err != nil {
+		s.Logger.WarnWithCtx(ctx, "Failed to remove log item from state file")
+	}
+}
```

Comment:
i think this should be defferred in case a return will occur?

Reply:
It depends - If an error occurs, we might want to remove the log item only from the in-memory state.
If we get here that means the pod finished streaming logs.
Perhaps only deferring the removal from the in-memory state then.